docs/source/dataframe-parquet.rst

```diff
@@ -0,0 +1,234 @@
+
+.. code-block:: python
+
+   >>> df = dd.read_parquet(
+   ...      "s3://bucket-name/my/parquet/",
+   ...      ignore_metadata_file=True  # don't read the _metadata file
+   ... )
+
+If no ``_metadata`` file is present, dask will load each parquet file
+individually as a partition in the larger dask dataframe (unless only a single
+file is specified). This is performant provided all files are of reasonable
+size (We recommend aiming for 10-250 MiB in-memory size per file once loaded
+into pandas). Too large files can lead to excessive memory usage on a single
+worker, while too small files can lead to poor performance as the overhead of
+dask dominates. If you need to read a parquet dataset composed of many large
+files (and lacking a ``_metadata`` file), you can pass
+``gather_statistics=True`` to have dask load the row-group metadata from all
+files in the dataset. Note that this can be *extremely* slow on some systems.
+
+Column Selection
+~~~~~~~~~~~~~~~~
+
+When loading parquet data, sometimes you don't need all the columns available
+in the dataset. In this case, you may want to specify the subset of columns you
+need via the ``columns`` kwarg. This is beneficial for two reasons:
+
+- It lets dask read less data from the backing filesystem, reducing IO costs
+- It lets dask load less data into memory, reducing memory usage
+
+.. code-block:: python
+
+    >>> dd.read_parquet(
+    ...     "s3://path/to/myparquet/",
+    ...     columns=["a", "b", "c"]  # Only read columns 'a', 'b', and 'c'
+    ... )
+
+
+Writing
+-------
+
+.. autosummary::
+    to_parquet
+    DataFrame.to_parquet
+
+Dask dataframe provides a `to_parquet` function (as well as a method) for
+writing parquet files and datasets.
+
+In its simplest usage, this takes a path to the directory in which to write the
+dataset. Just as with `read_parquet`, this path may be local, or point to some
+remote filesystem (e.g. S3, HDFS, ...) by prepending the path with a protocol.
+
+.. code-block:: python
+
+    # Write to a local directory
+    >>> df.to_parquet("path/to/my/parquet/")
+
+    # Write to S3
+    >>> df.to_parquet("s3://bucket-name/my/parquet/")
+
+Note that for remote filesystems you many need to configure credentials. When
+possible we recommend handling these external to Dask through
+filesystem-specific configuration files/environment variables (e.g. the
+``~/.aws/config`` file for S3). Alternatively, you can pass configuration on to
+the fsspec_ backend through the ``storage_options`` kwarg:
+
+.. code-block:: python
+
+   >>> df.to_parquet(
+   ...     "s3://bucket-name/my/parquet/",
+   ...     storage_options={'anon': True}  # passed to `s3fs.S3FileSystem`
+   ... )
+
+Dask will write one file per dask dataframe partition to this directory. To
+optimize access for downstream consumers, we recommend aiming for an in-memory
+size of 10-250 MiB per partition. This helps balance worker memory usage
```

Comment:
I'm not sure how general this memory size can be, given the existence of extremely large-memory machines, and the ability of both backends to write multiple row-groups per partition. The original ~100MB on-disk recommendation came out of balancing remote (S3) latency versus throughput, and happened to be the same order or magnitude as the more general guide to "have several times the in-memory footprint" for workers with ~4GB and a couple of threads.

Reply:
I agree that there are cases where this recommendation isn't perfect, but should be a decent general recommendation. I'm trying to balance a few factors here:

- Maximize IO performance/minimize extraneous IO
- Ensure the in-memory size of any partition isn't too large for common 4-8GB workers
- Ensure that downstream computations have enough data to work that dask's overhead is negligible.

I did mildly pick some arbitrary bounds though - if you have suggestions for adjusting these or the language around them that'd be useful. I'm trying to give good general recommendations to the majority of users, understanding the caveat that some users will always want to do something different.

Reply:
If you don't mind expanding the section, perhaps it's nice to be explicit? I mean, like saying: "assuming a latency of X and throughput of Y, it would be reasonable to aim for an on-disc partition size of Z"; and "following the guideline of having your dataset take up ~1/8 of your RAM/worker, a typical 4GB configuration would have you aiming for an in-memory size 250MB of per partition".

I would also add additionally something like: "the on disk size of your data versus in-memory varies a lot depending on factors such as the compression algorithm, so we recommend you estimate the factor for a subset of your data".

Reply:
I wouldn't worry about covering it in this PR, but it may also make sense to add a section on GPU considerations. On the RAPIDS side, we often encourage people to go for multi-row-group files where each **row-group** is ~128MB :)

Regardless of the hardware, however, it is typically the row-group size (rather than the partition/file size) that is most important on the read side. (Note that I didn't give this PR a full read yet, so that may already be covered)